felix/bpf/ut/bpf_prog_test.go

```diff
@@ -864,189 +893,164 @@ var udpDefault = &layers.UDP{
 	DstPort: 5678,
 }
 
-func testPacket(ethAlt *layers.Ethernet, ipv4Alt *layers.IPv4, l4Alt gopacket.Layer, payloadAlt []byte) (
+func testPacket(eth *layers.Ethernet, ipv4 *layers.IPv4, l4 gopacket.Layer, payload []byte) (
 	*layers.Ethernet, *layers.IPv4, gopacket.Layer, []byte, []byte, error) {
-	eth, ipv4, _, l4, bytes, payload, err := testPacket46(ethAlt, ipv4Alt, nil, l4Alt, payloadAlt, false)
-	return eth, ipv4, l4, bytes, payload, err
+	pkt := Packet{
+		eth:     eth,
+		ipv4:    ipv4,
+		l4:      l4,
+		payload: payload,
+		isIPv6:  false,
+	}
+	err := pkt.Generate()
+	return pkt.eth, pkt.ipv4, pkt.l4, pkt.payload, pkt.bytes, err
 }
 
-func testPacketv6(ethAlt *layers.Ethernet, ipv6Alt *layers.IPv6, l4Alt gopacket.Layer, payloadAlt []byte) (
-	*layers.Ethernet, *layers.IPv6, gopacket.Layer, []byte, []byte, error) {
-	eth, _, ipv6, l4, bytes, payload, err := testPacket46(ethAlt, nil, ipv6Alt, l4Alt, payloadAlt, true)
-	return eth, ipv6, l4, bytes, payload, err
+type Packet struct {
+	eth      *layers.Ethernet
+	ipv4     *layers.IPv4
+	ipv6     *layers.IPv6
+	l4       gopacket.Layer
+	udp      *layers.UDP
+	tcp      *layers.TCP
+	icmp     *layers.ICMPv4
+	icmpv6   *layers.ICMPv6
+	payload  []byte
+	bytes    []byte
+	layers   []gopacket.SerializableLayer
+	length   int
+	protocol layers.IPProtocol
+	isIPv6   bool
 }
 
-func testPacket46(ethAlt *layers.Ethernet, ipv4Alt *layers.IPv4, ipv6Alt *layers.IPv6, l4Alt gopacket.Layer, payloadAlt []byte, ipv6Enabled bool) (
-	*layers.Ethernet, *layers.IPv4, *layers.IPv6, gopacket.Layer, []byte, []byte, error) {
+func (pkt *Packet) handlePayload() {
+	if pkt.payload == nil {
+		pkt.payload = payloadDefault
+	}
+	pkt.length = len(pkt.payload)
+	pkt.layers = []gopacket.SerializableLayer{gopacket.Payload(pkt.payload)}
+}
 
-	var (
-		eth     *layers.Ethernet
-		ipv4    *layers.IPv4
-		ipv6    *layers.IPv6
-		payload []byte
-	)
+func (pkt *Packet) handleL4() error {
+	if pkt.l4 == nil {
+		pkt.l4 = udpDefault
+	}
 
-	if ethAlt != nil {
-		eth = ethAlt
-	} else {
-		if ipv6Enabled {
-			eth = ethDefaultWithIPv6
-		} else {
-			eth = ethDefault
-		}
+	switch v := pkt.l4.(type) {
+	case *layers.UDP:
+		pkt.udp = v
+		pkt.protocol = layers.IPProtocolUDP
+		pkt.layers = append(pkt.layers, pkt.udp)
+	case *layers.TCP:
+		pkt.tcp = v
+		pkt.protocol = layers.IPProtocolTCP
+		pkt.layers = append(pkt.layers, pkt.tcp)
+	case *layers.ICMPv4:
+		pkt.icmp = v
+		pkt.protocol = layers.IPProtocolICMPv4
+		pkt.layers = append(pkt.layers, pkt.icmp)
+	case *layers.ICMPv6:
+		pkt.icmpv6 = v
+		pkt.protocol = layers.IPProtocolICMPv6
+		pkt.layers = append(pkt.layers, pkt.icmpv6)
+	default:
+		return errors.Errorf("unrecognized l4 layer type %t", pkt.l4)
 	}
 
-	if ipv6Enabled {
-		if ipv6Alt != nil {
-			ipv6 = ipv6Alt
+	pkt.length += 8
+	return nil
+}
+
+func (pkt *Packet) setChecksum() {
+	switch pkt.protocol {
+	case layers.IPProtocolUDP:
+		if pkt.isIPv6 {
```

Comment:
Do you need the `isIPv6` You could have a method of  Packet.L3() that would return the L3 layer which ever is set or panic if both are set.

Reply:
removed it.